### PR TITLE
New hlsl colorspace

### PR DESCRIPTION
**Pull request by Przemog1**

### CODE REVIEW

File: include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl [hunk 1]
```diff
@@ -8,7 +8,10 @@
 #define NBL_CONSTEXPR constexpr
 #define NBL_CONSTEXPR_STATIC_INLINE constexpr static inline
 
-namespace nbl::hlsl
+namespace nbl
+{
+
+namespace hlsl
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl:14]: please undo this

File: include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl [hunk 2]
```diff
@@ -17,6 +20,15 @@ using add_reference = std::add_lvalue_reference<T>;
 template<typename T>
 using add_pointer = std::add_pointer<T>;
 
+// TODO[Przemog]: is there smarter way to do that? 
+// TODO[Przemog]: move functions below to other file (in cpp_compat drectory?)
+template<typename T, typename U>
+inline T lerp(const T& lhs, const T& rhs, const U& t)
+{
+    return glm::mix(lhs, rhs, t);
+}
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl:29]: `lerp` is a HLSL intrinsic, so lets move all the intrinsics implementations for C++ to `cpp_compat/intrinsics.h`

P.S. also make sure to explicitly specify the template parameters when calling `glm::` functions

File: include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl [hunk 3]
```diff
@@ -52,4 +65,12 @@ struct add_pointer
 
 #endif
 
+#include <nbl/builtin/hlsl/cpp_compat/promote.hlsl>
+
+#ifndef __HLSL_VERSION
+#define TYPENAME_CPP_ONLY typename
+#else
+#define TYPENAME_CPP_ONLY
+#endif
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl:74]: no, put the `typename` in the C++ versions of `..._REF_ARG`

File: include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl [hunk 4]
```diff
@@ -1,16 +1,11 @@
...
-#include "glm/detail/_swizzle.hpp"
-#include <stdint.h>
-#endif
-
 #include <nbl/builtin/hlsl/cpp_compat/vector.hlsl>
 
-namespace nbl::hlsl
+namespace nbl
+{
+namespace hlsl
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl:8]: this whole block was supposed to be under `ifndef __HLSL_VERSION`

File: include/nbl/builtin/hlsl/cpp_compat/vector.hlsl [hunk 5]
```diff
@@ -15,9 +22,73 @@ using vector = glm::vec<N, T>;
...
+
+using uint4 = vector<uint32_t, 4>;
+using uint3 = vector<uint32_t, 3>;
+using uint2 = vector<uint32_t, 2>;
+using uint1 = vector<uint32_t, 1>;
+
+using bool4 = vector<bool, 4>;
+using bool3 = vector<bool, 3>;
+using bool2 = vector<bool, 2>;
+using bool1 = vector<bool, 1>;
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/vector.hlsl:40]: does `float1` and stuff even exist in HLSL ?

Comment (Przemog1) [reply include/nbl/builtin/hlsl/cpp_compat/vector.hlsl:40]: yes

File: include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl (hunk 3 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl:74]: ok I fix it myself, just take my `cpp_compat` and nuke this one

File: include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl (hunk 1 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl:14]: jsut take my `cpp_compat.hlsl`

File: include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl (hunk 2 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl:29]: ok, nvm I see that `glm::` functions have vastly different template parameters

File: include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl (hunk 4 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl:8]: fixed it myself

File: include/nbl/builtin/hlsl/cpp_compat/vector.hlsl (hunk 5 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/cpp_compat/vector.hlsl:40]: added myself, please take my changes

File: include/nbl/builtin/hlsl/cpp_compat/vector.hlsl [hunk 6]
```diff
@@ -15,9 +22,73 @@ using vector = glm::vec<N, T>;
...
+template<typename T>
+struct ValueType
+{
+    using type = void;
+};
+template<>
+struct ValueType<float3>
+{
+    using type = float;
+};
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/vector.hlsl:87]: lets pawn this off to @atkurtul and his #550 PR,

we can probably follow a similar pattern to `underlying_type` for enums, and have our `scalar_type` for vectors

```cpp

```

File: include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl (hunk 2 above)
Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl:29]: btw I made it myself on `master`

**Comment by devshgraphicsprogramming** [comment]
merge `master` into this

File: include/nbl/builtin/hlsl/cpp_compat/vector.hlsl [hunk 7]
```diff
@@ -62,33 +67,25 @@ glm::vec<N, bool> operator>=(const glm::vec<N, T>& lhs, const glm::vec<N, T>& rh
...
 
-#endif
-
-}
+template<typename T, uint16_t N>
+struct scalar_type<vector<T,N> >
+{
+    using type = T;
+};
+}
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/vector.hlsl:88]: you want to create a `type_traits.hlsl` and slap that there along with specializations

File: include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl [hunk 8]
```diff
@@ -32,37 +31,100 @@ struct matrix final : private glm::mat<N,M,T>
...
+{
+
+//template<typename T, uint16_t N, uint16_t M>
+//struct scalar_type<matrix<T,N,M> >
+//{
+//    using type = T;
+//};
+
 }
 }
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl:128]: https://github.com/Devsh-Graphics-Programming/Nabla/pull/557/files/cd282ea4a3a067384169e2e87e9a8e73d3b28a2b..1f4e7725cadb0a752b156b063e7c72675d09e2f3#r1344743599

File: include/nbl/builtin/hlsl/cpp_compat/vector.hlsl (hunk 7 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/cpp_compat/vector.hlsl]: done

File: include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl (hunk 8 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/cpp_compat/matrix.hlsl]: done

File: include/nbl/builtin/hlsl/cpp_compat.hlsl [hunk 9]
```diff
@@ -9,7 +9,10 @@
 #define NBL_CONSTEXPR constexpr
 #define NBL_CONSTEXPR_STATIC_INLINE constexpr static inline
 
-namespace nbl::hlsl
+namespace nbl
+{
+
+namespace hlsl
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat.hlsl:15]: undo this and I merge

File: include/nbl/builtin/hlsl/cpp_compat.hlsl [hunk 10]
```diff
@@ -18,6 +21,13 @@ using add_reference = std::add_lvalue_reference<T>;
 template<typename T>
 using add_pointer = std::add_pointer<T>;
 
+//template<typename V>
+//struct scalar_type
+//{
+//    using type = void;
+//};
+
+}
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/cpp_compat.hlsl:30]: can you nuke?

Comment (Przemog1) [reply include/nbl/builtin/hlsl/cpp_compat.hlsl]: done

File: include/nbl/builtin/hlsl/cpp_compat.hlsl (hunk 9 above)
Comment (Przemog1) [reply include/nbl/builtin/hlsl/cpp_compat.hlsl]: done

**Pull request merged**